src/main/java/net/minecraftforge/fmlclient/StartupMessageRenderer.java

```diff
@@ -0,0 +1,279 @@
+            
+            VAO = glGenVertexArrays();
+            dataBuffer = glGenBuffers();
+            elementBuffer = glGenBuffers();
+            
+            glBindVertexArray(VAO);
+            
+            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, elementBuffer);
+            
+            glBindBuffer(GL_ARRAY_BUFFER, dataBuffer);
+            glVertexAttribPointer(inputLocation, 2, GL_FLOAT, false, 16, 0);
+            glEnableVertexAttribArray(inputLocation);
+            
+            glBindVertexArray(0);
+            
+            glBufferData(GL_ARRAY_BUFFER, bufferLength, GL_DYNAMIC_DRAW);
+            charBuffer = MemoryUtil.memAlloc(bufferLength);
+            
+            glBindBuffer(GL_ARRAY_BUFFER, previousArrayBuffer);
+            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, previousElementBuffer);
+            glBindVertexArray(previousVAO);
+        }
+        
+        public void close() {
+            MemoryUtil.memFree(charBuffer);
+            
+            glDeleteBuffers(elementBuffer);
+            glDeleteBuffers(dataBuffer);
+            glDeleteVertexArrays(VAO);
+            
+            glDeleteProgram(program);
+        }
+        
+        public void render(float width, float height, float textSize, boolean isDarkMode) {
+            this.height = height;
+            
+            int previousVAO = glGetInteger(GL_VERTEX_ARRAY_BINDING);
+            int previousArrayBuffer = glGetInteger(GL_ARRAY_BUFFER_BINDING);
+            int previousElementBuffer = glGetInteger(GL_ELEMENT_ARRAY_BUFFER_BINDING);
+            
+            int previousProgram = glGetInteger(GL_CURRENT_PROGRAM);
+            
+            int blendEnabled = glGetInteger(GL_BLEND);
+            int blendSrcMode = glGetInteger(GL_BLEND_SRC_ALPHA);
+            int blendDstMode = glGetInteger(GL_BLEND_DST_ALPHA);
+            
+            glEnable(GL_BLEND);
+            glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+            
+            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, elementBuffer);
+            glBindBuffer(GL_ARRAY_BUFFER, dataBuffer);
+            
+            glBindVertexArray(VAO);
+            glUseProgram(program);
+            
+            {
+                glUniform2f(posMultiplierLocation, 2.0f / width, 2.0f / height);
+                // TODO: 21/06/12: handle text resizing better, this is ok, but its not perfect
+                float size = Math.min(width / 854, height / 480) * textSize;
+                glUniform1f(sizeLocation, size);
+                
+                List<Pair<Integer, StartupMessageManager.Message>> messages = StartupMessageManager.getMessages();
+                for (int i = 0; i < messages.size(); i++) {
+                    boolean nofade = i == 0;
+                    final Pair<Integer, StartupMessageManager.Message> pair = messages.get(i);
+                    final float fade = clamp((4000.0f - (float) pair.getLeft() - (i - 4) * 1000.0f) / 5000.0f, 0.0f, 1.0f);
```

Comment:
I would avoid computing this if `nofade` is true. It's not much of a time save, but at this point, might as well